### PR TITLE
Encode ADT flags rather than recreating them from scratch at decoding time

### PR DESCRIPTION
**Pull request by compiler-errors**
r? lcnr

### CODE REVIEW

File: compiler/rustc_middle/src/ty/adt.rs [hunk 1]
```diff
@@ -249,15 +249,25 @@ impl Into<DataTypeKind> for AdtKind {
...
+        repr: ReprOptions,
+    ) -> Self {
+        AdtDefData { did, variants, flags, repr }
+    }
+
     /// Creates a new `AdtDefData`.
     pub(super) fn new(
         tcx: TyCtxt<'_>,
-        did: DefId,
+        did: LocalDefId,
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/adt.rs:264]: Setting this to local so we enforce that `AdtDefData::new` is called only on local def ids.

File: compiler/rustc_middle/src/ty/adt.rs [hunk 2]
```diff
@@ -249,15 +249,25 @@ impl Into<DataTypeKind> for AdtKind {
 }
 
 impl AdtDefData {
+    pub(super) fn new_from_flags(
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/adt.rs:252]: this ctor is kinda redundant, i could also just make all the fields `pub(super)` -- *OR* just pull `TyCtxt::mk_adt_def_from_flags`/`TyCtxt::mk_adt_def` into a new impl in this module. I don't really care.

**Comment by compiler-errors** [comment]
Likely noise

**Comment by oli-obk** [comment]
@bors r-

The binary size regression is real

What's the motivation for this?

**Comment by Kobzol** [comment]
(Note that this only seems to affect metadata size, not actual binary size of binaries.)

I was also wondering about the motivation given this size regression though.

**Comment by compiler-errors** [comment]
This was requested in https://github.com/rust-lang/rust/pull/127437#discussion_r1667712319. @lcnr has some interest in migrating the compiler to stop encoding built-in attrs across crates. IDK if they have a tracking issue for this, but if they don't, they should probably add it. I don't particularly care about this work, so if anyone wants to pick it up, they can do it.

**Pull request closed without merging**